### ISSUE TITLE
Replace `defaultProps` in function components with default arguments

### ISSUE THREAD

**Issue by iamandrewluca** [strapi/strapi#18695]
React Team has an RFC in the future to deprecate `defaultProps` on Function Components in favour of default arguments. This should give a performance improvement because React does not need to do those checks under the hood to see if the Component has `defaultProps`.

This should also make migration to TypeScript easier in the future of JS Components, and maybe the removal of `propTypes` if they will not be needed anymore after TypeScript migration.

> **Note**
> These changes should target the v5 branch.

## [Motivation](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#motivation)

> We need to do a dynamic test against a component if it has a `defaultProps` during every element creation call. This can't be optimised well because the function it is called within is highly megamorphic.

> Other than performance this is also about a long-term simplification of the number of concepts you have to learn to use React. In particular, `forwardRef` and `defaultProps` are no longer something special.

> `defaultProps` in element creation doesn't work with `React.lazy` so in that case we also have to check for resolving `defaultProps` in the render phase, which means that the semantics are inconsistent anyway.

## [Deprecate `defaultProps` on function components](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#deprecate-defaultprops-on-function-components)

> In function components there really isn't much need for this pattern since you can just use JS default arguments and all the places where you typically use these values are within the same scope.

## Packages 

- [ ] `@strapi/admin` 142 instances
	- [ ] `content-manager` 72 instances
	- [ ] `pages` 57 instances
	- [ ] `ee` 13 instances
- [ ] `@strapi/plugin-content-type-builder` 31 instances
- [ ] `@strapi/plugin-upload` 56 instances
- [ ] `@strapi/plugin-i18n` 7 instances
- [ ] `@strapi/plugin-users-permissions` 4 instances

ps: I wanted to start working on these changes, but then I decided that maybe an issue and an approval would be better first.

**Comment by derrickmehaffy** [comment]
Hey hey, I'll let the engineers take a look but just an FYU we are on a v4 change freeze until March 2024 while the team works on v5.

I think it's best we wait for v5 before doing major reworking like this until then so we don't have to deal with all the merge conflicts between v4/v5.

This might also impact our design system repo too: https://GitHub.com/Strapi/design-system

**Comment by iamandrewluca** [comment]
Hey hey, I heard about v4 freeze. I was thinking of these changes to be done directly in v5. To avoid conflicts, multiple PRs can be done as I grouped in the PR description, by plugin. Anyway, if there is a freeze for external contributors on v5, we can wait until March.

In the design-system repo is using fully TypeScript and does not use `defaultProps` and `propTypes`

**Comment by joshuaellis** [comment]
Hey @iamandrewluca, thanks for sharing this – we're actually converting the whole codebase to Typescript so we'll be getting rid of `defaultProps` anyway & moving to default arguments as the post suggests. As Derrick said we're on a community freeze so we won't be accepting any contributions as a blanket rule unfortunately. When we've merged v5 into `main` feel free to check back if you're still interested in contributing as there may be some parts of the codebase we've had to stall the migration on!
